compiler/src/iree/compiler/Codegen/Common/CPU/CPUPrepareUkernels.cpp

```diff
@@ -277,6 +312,83 @@ struct Convert3DPackto2DPackPattern : public OpRewritePattern<tensor::PackOp> {
   }
 };
 
+struct Convert5DUnPackto4DUnPackPattern
+    : public OpRewritePattern<tensor::UnPackOp> {
+  using OpRewritePattern<tensor::UnPackOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::UnPackOp unpackOp,
+                                PatternRewriter &rewriter) const override {
+    if (unpackOp.getSourceRank() != 5 || unpackOp.getDestRank() != 3) {
+      return failure();
+    }
+
+    llvm::SmallDenseSet<int64_t> s(unpackOp.getInnerDimsPos().begin(),
+                                   unpackOp.getInnerDimsPos().end());
+
+    SmallVector<int64_t> seqOrOuterDimsPerm =
+        llvm::to_vector(llvm::seq<int64_t>(0, unpackOp.getDestRank()));
+    if (!unpackOp.getOuterDimsPerm().empty()) {
+      applyPermutationToVector(seqOrOuterDimsPerm, unpackOp.getOuterDimsPerm());
+    }
+
+    int64_t srcPos, destPos = 0;
+
+    for (auto [idx, val] : llvm::enumerate(seqOrOuterDimsPerm)) {
+      if (s.contains(val))
+        continue;
+      srcPos = idx;
+      destPos = val;
+      break;
+    }
```

Comment:
Is it possible to refactor this out and use it for pack patterns as well? tensor.pack op and tensor.unpack op have many same methods by design (e.g., getOuterDimsPerm(), getInnerDimsPos(), etc)`, so I think we can use a template.

Reply:
It's better to have them separately. A user chiming in to make changes to the `pack` op pattern will have to think about the `unpack` op, too, and things may go wrong, creating a dev overhead. I can create functions that can be used by both transformations. Also, there would be conditionals involved. Does this sound good to you? Otherwise, I can add templates, too.